### PR TITLE
datetime: test RFC 3339 and ISO 8601 formats

### PR DESCRIPTION
**Pull request by ligurio**
The commit 3c40366172e3 ("datetime, lua: date parsing functions") introduced a function `datetime.parse()`. The function accepts a datetime format: 'iso8601', 'rfc3339' and a custom datetime format. However, 'iso8601' as well as 'rfc3339' is a set of datetime formats (see [1] and [2]), and `datetime.parse()` supports a subset of these formats. There is a documentation issue [3], where we should describe what formats are supported and what formats are not. Supported formats must be added to the testing and the patch do it.

1. https://datatracker.ietf.org/doc/html/rfc3339
2. https://ijmacd.github.io/rfc3339-iso8601/
3. https://github.com/tarantool/doc/issues/3155

Needed for tarantool/doc#3155

NO_CHANGELOG=testing
NO_DOC=gh-3155

### CODE REVIEW

**Comment by ligurio** [comment]
@Buristan I've updated a test as you suggested and added comments to split datetimes by types.

**Review by Buristan — APPROVED** [review]
Hi, Sergey!
Thanks for the decent work!
LGTM, with a few minor nits below.

> and what formats are not. Supported formats must be added to
the testing and the patch do it.

I would rephrase this message like the following:
```
Patch adds a tests for most of supported formats.
```
Also, as you mentioned offline, the set of formats is not exhaustive, see [1]. Should we add more tests?

[1]: https://github.com/IJMacD/rfc3339-iso8601?tab=readme-ov-file#format-listing

File: test/app-luatest/datetime_test.lua [hunk 1]
```diff
@@ -0,0 +1,2101 @@
...
+            fmt = '%Y-%M-%DT%h:%m:%s+00:00',
+            buf = '2024-07-31T14:30:02+00:00',
+        }, {
+            fmt = '%Y-%M-%DT%h:%m:%.3s+00:00',
+            buf = '2024-07-31T14:30:02.132+00:00',
+        },
+    },
+
+    ['RFC3339 ONLY'] = {
+        -- Dates.
```
[the remark this answers is not in the document]
Comment (ligurio) [reply test/app-luatest/datetime_test.lua:49]: Right, fixed.

File: test/app-luatest/datetime_test.lua [hunk 2]
```diff
@@ -0,0 +1,2101 @@
...
+        },
+    }
+
+}
+
+local pg = t.group('pgroup')
+
+-- XXX: It is not possible to use parameterization by passing a
+-- table with test parameters to `t.group` because datetime format
+-- strings in test parameters contains symbol `/` that is not
```
[the remark this answers is not in the document]
Comment (ligurio) [reply test/app-luatest/datetime_test.lua:2043]: Fixed.

```diff
@@ -2040,7 +2044,7 @@ local pg = t.group('pgroup')
 
 -- XXX: It is not possible to use parameterization by passing a
 -- table with test parameters to `t.group` because datetime format
--- strings in test parameters contains symbol `/` that is not
+-- strings in test parameters contains the symbol `/` that is not
 -- allowed in testcases names. Source code below inserts
 -- test functions to a test group with testnames where `/` is
 -- replaced with `_`.
```

File: test/app-luatest/datetime_test.lua [hunk 3]
```diff
@@ -0,0 +1,2101 @@
...
+-- allowed in testcases names. Source code below inserts
```
[the remark this answers is not in the document]
Comment (ligurio) [reply test/app-luatest/datetime_test.lua:2044]: Fixed.

File: test/app-luatest/datetime_test.lua [hunk 4]
```diff
@@ -0,0 +1,2101 @@
...
+-- test functions to a test group with testnames where `/` is
```
[the remark this answers is not in the document]
Comment (ligurio) [reply test/app-luatest/datetime_test.lua:2045]: Fixed.

File: test/app-luatest/datetime_test.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (ligurio) [reply test/app-luatest/datetime_test.lua:2044]:
```diff
@@ -2040,9 +2044,9 @@ local pg = t.group('pgroup')
 
 -- XXX: It is not possible to use parameterization by passing a
 -- table with test parameters to `t.group` because datetime format
--- strings in test parameters contains symbol `/` that is not
--- allowed in testcases names. Source code below inserts
--- test functions to a test group with testnames where `/` is
+-- strings in test parameters contains the symbol `/` that is not
+-- allowed in testcases names. The source code below inserts
+-- test functions into a test group with testnames where `/` is
 -- replaced with `_`.
 for supported_by, standard_cases in pairs(SUPPORTED_DATETIME_FORMATS) do
     for _, case in ipairs(standard_cases) do
``

File: test/app-luatest/datetime_test.lua [hunk 5]
```diff
@@ -0,0 +1,2101 @@
+local t = require('luatest')
+local dt = require('datetime')
+
```
[the remark this answers is not in the document]
Comment (ligurio) [reply test/app-luatest/datetime_test.lua:3]: Added a paragraph to the commit message.

**Comment by ligurio** [comment]
@Buristan 

> Also, as you mentioned offline, the set of formats is not exhaustive, see [^1]. Should we add more tests?

Yes, we should, but it is better to do it with fuzzing because exhaustive testing of formats defined by grammar by only example-based tests is complicated. The initial idea was to describe supported formats in documentation and have a good regression testsuite for `strptime`. I suppose the goal is reached. Also, I've a fuzzing test for datetime datatype and someday I'll merge it to master.

[^1]: https://github.com/IJMacD/rfc3339-iso8601?tab=readme-ov-file#format-listing

**Review by ImeevMA — APPROVED** [review]
